macros/shared/src/test/scala-2.x/zio/macros/AccessibleMMSpec.scala

```diff
@@ -0,0 +1,385 @@
+        assertM(typeCheck {
+          """
+            @accessibleMM[RManaged]
+            object Module {
+              trait Service[F[_, _]] {
+                def managed(s: String): F[Has[Unit], Int]
+              }
+            }
+
+            object Check {
+              def managed(s: String): ZManaged[Has[Module.Service[RManaged]] with Has[Unit], Throwable, Int] =
+                Module.managed(s)
+            }
+          """
+        })(isRight(anything))
+      },
+      testM("generates accessor for service with default method implementations") {
+        assertM(typeCheck {
+          """
+             @accessibleMM[IO]
+             object Module {
+               trait Service[F[_, _]] {
+                 def foo(x: Int): F[Unit, Unit] = foo(x.toString)
+                 def foo(x: String): F[Unit, Unit]
+               }
+             }
+
+             object Check {
+              def foo(x: Int): ZIO[Has[Module.Service[IO]], Unit, Unit] =
+                Module.foo(x)
+              def foo(x: String): ZIO[Has[Module.Service[IO]], Unit, Unit] =
+                Module.foo(x)
+            }
+          """.stripMargin
+        })(isRight(anything))
+      },
+      testM("generates accessor for service with one type param other than F") {
+        assertM(typeCheck {
+          """
+             @accessibleMM[IO]
+             object Module {
+               trait Service[F[_, _], T] {
+                 val v: F[Int, T]
+                 def f1: F[Int, Unit]
+                 def f2(): F[Int, Unit]
+                 def f3(t: T): F[Int, Unit]
+                 def f4(t: T)(i: Int): F[Int, Unit]
+                 def f5(t: T)(implicit i: Int): F[Int, Unit]
+                 def f6(t: T*): F[Int, Unit]
+               }
+             }
+
+             object Check {
+               def v[T: Tag]: ZIO[Has[Module.Service[IO, T]], Int, T] =
+                 Module.v[T]
+               def f1[T: Tag]: ZIO[Has[Module.Service[IO, T]], Int, Unit] =
+                 Module.f1[T]
+               def f2[T: Tag](): ZIO[Has[Module.Service[IO, T]], Int, Unit] =
+                 Module.f2[T]()
+               def f3[T: Tag](t: T): ZIO[Has[Module.Service[IO, T]], Int, Unit] =
+                 Module.f3[T](t)
+               def f4[T: Tag](t: T)(i: Int): ZIO[Has[Module.Service[IO, T]], Int, Unit] =
+                 Module.f4[T](t)(i)
+               def f5[T: Tag](t: T)(implicit i: Int): ZIO[Has[Module.Service[IO, T]], Int, Unit] =
+                 Module.f5[T](t)
+               def f6[T: Tag](t: T*): ZIO[Has[Module.Service[IO, T]], Int, Unit] =
+                 Module.f6[T](t: _*)
+             }
+          """
+        })(isRight(anything))
+      },
+      testM("generates accessor for service with contravariant type param") {
+        assertM(typeCheck {
+          """
+             @accessibleMM[RIO]
+             object Module {
+               trait Service[-A, F[_, _]] {
```

Comment:
It would be good to add tests also for variance annotations inside the F parameter as in:

```scala
trait Service[F[+_, _]] // should work and variance should be preserved in output code (e.g. Service[UIO] <: Service[Task])
trait Service[F[_, +_]] // should work and variance should be preserved in output code (e.g. Service[UIO] <: Service[Task])
trait Service[F[+_, +_]] // should work and variance should be preserved in output code (e.g. Service[UIO] <: Service[Task])

trait Service[F[-_, _]] // should fail because IO is not contravariant
trait Service[F[_, -_]] // should fail because IO is not contravariant
```

Reply:
I've added 4 more test at the end of `AccessibleMMSpec` and `AccessibleMSpec`. They don't cover all cases due to limitations of typechecking in tests, but I hope this is better than nothing.
Thank you for your help!